spmd/tensor/ops/prop_rules.py

```diff
@@ -0,0 +1,120 @@
+# Copyright (c) Meta Platforms, Inc. and affiliates
+from typing import Tuple, List, Dict, Optional
+from spmd.tensor.placement_types import _Partial, PlacementSpec
+
+
+def _parse_einop_equation(equation: str) -> Tuple[List[str], List[str]]:
+    inputs, outputs = equation.split("->")
+    input_dims = []
+    output_dims = []
+    for input_dim in inputs.split(","):
+        input_dims.append(input_dim)
+
+    for output_dim in outputs.split(","):
+        output_dims.append(output_dim)
+
+    return input_dims, output_dims
+
+
+def einop_prop(
+    equation: str, input_specs: Tuple[PlacementSpec, ...], linear=False
+):
+    """
+    Propagate the sharding of inputs to output for ops whose data
+    moves according to einsum notation. This is mostly borrowed
+    from @zdevito's sharding simulator. Examples:
+        mk,kn->mn - einsum
+        ij,ij->ij - addition
+        ij,j->ij - broadcasted addition
+        ij->i - reduction
+    Other ops could use this propagation algorithm when applied.
+    """
+    input_dims, output_dims = _parse_einop_equation(equation)
+    # NOTE: only support single output unless needed in future
+    output_dim = output_dims[0]
+    dim_to_sharding: Dict[str, int] = {}
+    pending_sums: List[int] = []
+
+    seen_shardings = {}
+
+    for input in input_specs:
+        for i, a in enumerate(input.placements):
+            if a.is_partial():
+                seen_shardings[i] = "+"
+                if not i in pending_sums:
+                    pending_sums.append(i)
+                if not linear:
+                    raise RuntimeError(
+                        "cannot do generic op on a tensor with partial sums"
+                    )
+
+    for input_dim, input_spec in zip(input_dims, input_specs):
+        for dim, mesh_dim in zip(input_dim, input_spec.dim_map):
+            if dim not in dim_to_sharding:
+                dim_to_sharding[dim] = mesh_dim
+            else:
+                # TODO: merge the sharding properly
+                assert dim_to_sharding[dim] == mesh_dim, ""
+
+    for dim, shard_on_mesh in dim_to_sharding.items():
+        if dim not in output_dims[0] and shard_on_mesh != -1:
+            pending_sums.append(shard_on_mesh)
+
+    return PlacementSpec.from_dim_map(
+        input_specs[0].mesh,
+        [dim_to_sharding[dim] for dim in output_dim],
+        pending_sums,
+    )
+
+
+def mm_prop(
+    mat1_spec: PlacementSpec, mat2_spec: PlacementSpec
+) -> Optional[PlacementSpec]:
```

Comment:
I think it's not hard to generalize this one , maybe you can do it here:

1) group sharding by dimension of the tensor.
2) check that the sharding-mesh-dim of dimension=1 of mat1 is the same as sharding of dimension=0 of mat2 (or that they are both replicated).
3) check that the sharding for dim=0 of mat1 and dim=1 of mat2 are _not_ on the same mesh dimension as the other ones above.

Reply:
yeah this make sense, this PR is getting too long, let me create a follow up issue and we can update incrementally :)

Reply:
https://github.com/pytorch/PiPPy/issues/335